### PR TITLE
Core: Add typed dictionary support for binary serialization

### PR DESCRIPTION
**Pull request by dalexeev**
* This was skipped in #78656.
* See also #78219 and subsequent PRs.

### CODE REVIEW

**Review by Repiteo — APPROVED** [review]
Everything seems to be in order!

File: tests/core/io/test_marshalls.h [hunk 1]
```diff
@@ -386,6 +386,89 @@ TEST_CASE("[Marshalls] Typed array decoding") {
...
 
+TEST_CASE("[Marshalls] Typed dicttionary encoding") {
+	int r_len;
+	Dictionary dictionary;
+	dictionary.set_typed(Variant::INT, StringName(), Ref<Script>(), Variant::INT, StringName(), Ref<Script>());
+	dictionary[Variant(uint64_t(0x0f123456789abcdef))] = Variant(uint64_t(0x0f123456789abcdef));
+	uint8_t buffer[40];
+
+	CHECK(encode_variant(dictionary, buffer, r_len) == OK);
+	CHECK_MESSAGE(r_len == 40, "Length == 4 bytes for header + 8 bytes for dictionary type + 4 bytes for dictionary size + 24 bytes for key-value pair");
```
Comment (AThousandShips) [inline tests/core/io/test_marshalls.h:397]:
```suggestion
	CHECK_MESSAGE(r_len == 40, "Length == 4 bytes for header + 8 bytes for dictionary type + 4 bytes for dictionary size + 24 bytes for key-value pair.");
```

**Review by AThousandShips — APPROVED** [review]

File: core/io/marshalls.cpp [hunk 2]
```diff
@@ -126,7 +201,9 @@ Error decode_variant(Variant &r_variant, const uint8_t *p_buffer, int p_len, int
 		*r_len = 4;
 	}
 
-	// Note: We cannot use sizeof(real_t) for decoding, in case a different size is encoded.
+#define GET_CONTAINER_TYPE_KIND(m_id) ((ContainerTypeKind)((header & HEADER_DATA_FIELD_TYPED_##m_id##_MASK) >> HEADER_DATA_FIELD_TYPED_##m_id##_SHIFT))
+
```
Comment (Faless) [inline core/io/marshalls.cpp:205]: Should be moved close to other CONTAINER_TYPE defines

File: core/io/marshalls.cpp [hunk 3]
```diff
@@ -1250,6 +1303,8 @@ Error decode_variant(Variant &r_variant, const uint8_t *p_buffer, int p_len, int
 		}
 	}
 
+#undef GET_CONTAINER_TYPE_KIND
+
```
Comment (Faless) [inline core/io/marshalls.cpp:1307]: no need to undef

**Review by Faless — COMMENTED** [review]
I'll need to test and try to do a proper re-review since it's hard to safely evaluate just by looking at the changes, but in the meantime I added a couple of readability comments

**Review by Faless — APPROVED** [review]
Looking great!

I've gone through all changes, everything seems in order, also run some tests using the multiplayer API (which adds some custom encoding on top of the regular variant marshalling) and everything worked as expected.

Awesome work! :1st_place_medal: :trophy:

**Comment by Repiteo** [comment]
Thanks!

**Pull request merged**